### ISSUE TITLE
Question:  Autotools flags and patches for projects targeting Android + Project dependencies with versions

### ISSUE THREAD

**Issue by pretyman** [ruslo/hunter#149]
Hi Ruslan,

First, thank you so much for hunter!

Question 1) I want to add a couple of projects to hunter based on autotools. I am following the OpenSSL project as a template. Today I use ExternalProject_Add with a CONFIGURE_COMMAND that changes if I am compiling for Mac or Android. 

When compiling for android I must add after the ./configure parameters: CPP CPPFLAGS,CC, CFLAGS, CXX, CXXFLAGS, LDFLAGS, RANLIB, STRIP, LD, OBJDUMP, NM and LIBS. 
CPPFLAGS contains " -I${ANDROID_SYSROOT}/usr/include "
while LDFLAGS would contain " -L@HUNTER_PACKAGE_INSTALL_PREFIX@/lib "

Sometime I need to reweite the whole ./configure commandline, which is a sign of not a very good design of by build cmake scripts.

My Question is: where would be the correct place to place these definitions? They get re-used for every other android project based on autotools, except sometimes I need to add a few extra LDFLAGS on a per project basis. Also sometimes a project will need a configure parameter as " --with-projectA=@HUNTER_PACKAGE_INSTALL_PREFIX@/bin/other_dependency " but then I believe I should create a scheme file for that. Correct?

Question 2) Most projects I must apply a patch to compile with Android, this is done a per project basis. Should I have a scheme file for each of these projects? I use the patch command for that, so I believe this wouldn't be portable to M$ platforms.

Question 3 - low priority) If I have a package B version 4.0.0 that depends on version 3.4.x of package A, is there anyway to express that? Maybe I should clearly define that B v4.0.0 depends on A v3.4.2. I saw that LLVM depends on Clang as per ${HUNTER_ROOT}/hunter/cmake/projects/LLVM/hunter.cmake

hunter_add_package(Clang) # CLANG_ROOT
hunter_add_package(ClangToolsExtra) # CLANGTOOLSEXTRA_ROOT
...

but then hunter_add_version is called with the version of LLVM, never specifying which Clang version (although in this specific example they way hand in hand with versions, other libraries don't).

Thank you once again, I hope I didn't sound confusing as I just woke up.

**Comment by ruslo** [comment]
> that changes if I am compiling for Mac or Android

Just for your information I haven't test Android yet but it's next milestone: https://github.com/ruslo/hunter/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.11

**Comment by ruslo** [comment]
> Sometime I need to reweite the whole ./configure commandline, which is a sign of not a very good design of by build cmake scripts.

What about using [hunter_pick_scheme](https://github.com/ruslo/hunter/wiki/dev.modules#hunter_pick_scheme)? I guess we can add `ANDROID` just like `IPHONEOS`

**Comment by ruslo** [comment]
> where would be the correct place to place these definitions?

My first thought is that all the global flags that is not related to the specific package must go to the toolchain. See https://github.com/ruslo/polly. Also I'm planning to add this collection to Hunter as a package, so we don't have to download it manually: https://github.com/ruslo/hunter/issues/150

**Comment by ruslo** [comment]
> Also sometimes a project will need a configure parameter as " --with-projectA=@HUNTER_PACKAGE_INSTALL_PREFIX@/bin/other_dependency " but then I believe I should create a scheme file for that. Correct?

Well yes, but we can think about "universal" autotool scheme. So you can set "--with-*" options using CMake arguments. Just like `boost.b2` build tool: https://github.com/ruslo/hunter/wiki/pkg.boost#cmake-options

I.e. `CMAKE_ARGS WITH_PROJECTA=...` add extra flag `--with-projectA=...` or something

**Comment by ruslo** [comment]
> I use the patch command for that, so I believe this wouldn't be portable to M$ platforms

I prefer to use separate repositories for it: https://github.com/hunter-packages
With github it's quite easy to create new one.

The reason is that it can be quite a lot of patch-work (different developers, different versions, ...) so we need to have some patch-orginizing system. Why not use git? :)

**Comment by ruslo** [comment]
> If I have a package B version 4.0.0 that depends on version 3.4.x of package A, is there anyway to express that?

All versions of packages calculates only once. From user's point of view it happens in `HunterGate` command. Versions of packages by default can be found in [cmake/configs/default](https://github.com/ruslo/hunter/blob/master/cmake/configs/default.cmake). And user can customize it: https://github.com/ruslo/hunter/wiki/example.custom.config.id#custom-package-version

**Comment by pretyman** [comment]
Hi Ruslan, thank you for the feedback.

I'm a bit of an idiot and I missed the Android branch of hunter, only saw now by your commit 26d9f2191f2bb360c260b86b3b02c750a3dba8e2

Now that I saw it, I ask if you have plans for "teaching" projects how to compile on the Android platform?

I did small changes to hunter_pick_scheme as shown on 8eac558fcfb6085d183a3f1148e6664229930823 that I am using to compile boost, but I am not sure how will that scale. 
What if later Raspberry Pi support would be added, would that mean that hunter_pick_scheme would need to changed?

The boost scheme used is at commit 2780ddf2b17dc11570f13fcc89d6e137f010bb5b, and I am unsure of 120 to 123, it looks that is not the right place to do that, since it would be re-used in other places, I would need to follow your DRY mantra, but I am unsure where to put it.

I also have autotools schemes at de2cc79ed52160e8f10ac67d194d1ca23e73a894 for default and android, to compile autotools based projects, but they seem messy and a little bit wrong, so I don't think you would be able to take advantage of them.

I am sorry I can't contribute much since I am a beginner at CMake, but I learning  a lot from hunter.

**Comment by ruslo** [comment]
> Now that I saw it, I ask if you have plans for "teaching" projects how to compile on the Android platform?

Currently I'm busy with it. I hope I will have something working about next two days. As far as I understand CMake-based projects will work without changes (at least libraries).

> I did small changes to hunter_pick_scheme as shown on 8eac558 that I am using to compile boost, but I am not sure how will that scale

Yep, something like that. But we need to check some variable that shows us that we are currently building on Android. May be [CMAKE_ANDROID_API](http://www.cmake.org/cmake/help/v3.2/variable/CMAKE_ANDROID_API.html)?

```cmake
string(COMPARE NOTEQUAL "${CMAKE_ANDROID_API}" "" is_android)
...
elseif(is_android AND x_ANDROID)
  set(HUNTER_DOWNLOAD_SCHEME "${x_ANDROID}")
...
```

Otherwise scheme from `ANDROID` will be picked by default (i.e. instead of `DEFAULT`).

> What if later Raspberry Pi support would be added, would that mean that hunter_pick_scheme would need to changed?

It depends on scheme. If it's not "universal" enough then yes. That's the curse of non-CMake builds.

> The boost scheme used is at commit 2780ddf, and I am unsure of 120 to 123, it looks that is not the right place to do that, since it would be re-used in other places, I would need to follow your DRY mantra, but I am unsure where to put it.

It looks quite the same as `url_sha1_boost_library.cmake.in` so may be we can just put some if's instead of making new scheme (?):
```cmake
if(is_android)
  set(
    build_opts
    toolset=gcc-android
    target-os=linux
    ${build_opts}
  )
endif()
```

> I also have autotools schemes at de2cc79 for default and android, to compile autotools based projects, but they seem messy and a little bit wrong, so I don't think you would be able to take advantage of them

Same here, probably it can be just:
```cmake
if(is_android)
  set(configure_opts ${configure_opts} --host=${ANDROID_TOOLCHAIN_MACHINE_NAME})
endif()
```

**Comment by ruslo** [comment]
> Yep, something like that. But we need to check some variable that shows us that we are currently building on Android. May be CMAKE_ANDROID_API?

I guess we can just check `ANDROID` variable: https://github.com/taka-no-me/android-cmake/blob/556cc14296c226f753a3778d99d8b60778b7df4f/android.toolchain.cmake#L1522

**Comment by pretyman** [comment]
For boost I updated the url_ as per https://github.com/pretyman/hunter/commit/5e41380ba5b63a73bd3bf225d530ec7e55e14eff - this is really messy, it is just to show you where I am stuck: I need to add some extra configuration to boost.user.jam: the toolset version and the archiver.

 using ${toolset_name} 
     : ${toolset_version} 
     : \"${boost_compiler}\"
     : <archiver> ${CMAKE_AR} 
     ;

I tried to set the whole content as a variable as boost_user_jam_contents  at https://github.com/pretyman/hunter/commit/5e41380ba5b63a73bd3bf225d530ec7e55e14eff#diff-72b4f26b69a795d0ed61fc4edbca8ad2R169 , however it only includes the first 3 lines and not the rest, creating a malformed boost.user.jam

Also not that the "toolset=" argument has to change to be "toolset=toolset_name-toolset_version" which is treated at https://github.com/pretyman/hunter/commit/5e41380ba5b63a73bd3bf225d530ec7e55e14eff#diff-72b4f26b69a795d0ed61fc4edbca8ad2R214 

However I could not compile it on the Mac since I had a python version problem, as the python version that cmake finds is 2.7 and the Mac needs a script that runs on python 3.x, failing the build when I tested it.

**Comment by ruslo** [comment]
@pretyman Let's move discussion here: [Android Boost](https://github.com/ruslo/hunter/issues/154)

**Comment by ruslo** [comment]
@pretyman Let's move `autotool` discussion here: [Android autotool](https://github.com/ruslo/hunter/issues/158)

**Comment by pretyman** [comment]
Ok, good idea to move the discussion to their proper places. I need to catch up with the recent changes made by you. I have 2 jobs and can only play with hunter/poly/android in the spare time and this means I will only have something to contribute by Sunday night.

**Comment by ruslo** [comment]
> will only have something to contribute by Sunday night

no problems, I'm planning to investigate how to create apk easily (I want to see working application first, not just libraries) and then will take a look at boost android
